generateBins.py

```diff
@@ -25,69 +13,50 @@
 
 
 # TODO Still need a good way of supplying these
-PFPR_FILE       = "rwa_pfpr2to10.asc"
+PFPR_FILE = "rwa_pfpr2to10.asc"
 POPULATION_FILE = "rwa_population.asc"
 
-
 # TODO Determine the bins computationally
 # Following bins are for Rwanda
 POPULATION_BINS = [2125, 5640, 8989, 12108, 15577, 20289, 27629, 49378, 95262, 286928]
 
-# getting data ready for binning
-# data should be 1-dimensional array, python list or iterable
-myArray  = np.loadtxt("GIS\\rwa_population.asc", skiprows = 6)
-print (type(myArray))
-print(myArray.ndim)
-print(myArray.shape)
-print(myArray.size)
-array_1d = myArray.flatten()
-print (array_1d)
-print(array_1d.ndim)
-print(array_1d.shape)
-print(array_1d.size)
-#gvf = 0.0
-#nclasses = 20
-
-def goodness_of_variance_fit(array, classes):
-    # get the break points
-    classes = jenkspy.jenks_breaks(array, classes)
-
-    # do the actual classification
-    classified = np.array([classify(i, classes) for i in array])
-
-    # max value of zones
-    maxz = max(classified)
-
-    # nested list of zone indices
-    zone_indices = [[idx for idx, val in enumerate(classified) if zone + 1 == val] for zone in range(maxz)]
-
-    # sum of squared deviations from array mean
-    sdam = np.sum((array - array.mean()) ** 2)
-
-    # sorted polygon stats
-    array_sort = [np.array([array[index] for index in zone]) for zone in zone_indices]
-
-    # sum of squared deviations of class means
-    sdcm = sum([np.sum((classified - classified.mean()) ** 2) for classified in array_sort])
-
-    # goodness of variance fit
-    gvf = (sdam - sdcm) / sdam
-
-    return gvf
 
-
-def process(configuration, gisPath = ""):
+def process(configuration, gisPath=""):
     # Load the configuration
     cfg = load_configuration(configuration)
+    filename = os.path.join(gisPath, PFPR_FILE)
 
     # TODO Add the stuff for the population bins!
+    # GVF Implementation
+    binning = bin_asc(filename)
+
+    list_bins = []
+    list_GVF = []
+    # Assuming the range for bins to be between 5 and 30 (computing the optimal number of bins w.r.t. GVF), to avoid overfitting or underfitting of data
+    for x in range(5, 30):
+        GVF = goodness_of_variance_fit(binning, x)
+        # GVF less than 0.7 might not be the best fit for data; GVF ~ 0 (No fit), GVF ~ 1 (Perfect fit)
+        if GVF < 0.7:
+            print ("Bins=", x,"GVF = ", GVF, "Warning: GVF too low")
+        # Storing the values of bins and GVF in a list
+        list_bins.append(x)
+        list_GVF.append(GVF)
+    # computing difference between GVF values and storing them as (key, value) pairs in dictionary
+    diff_GVF = {list_bins[i]: abs(list_GVF[i] - list_GVF[i - 1]) for i in range(len(list_bins))}
```

Comment:
You're on the right track here, but you can find the while going through each bin size in the range, then just break out of the for loop once the conditions is satisfied. Since we want to use the smallest number of bins, using the previous bin size should be fine. That also makes things a bit interesting if we reach the end of the loop without the heuristic being meet, it's up to you how to handle that.